### PR TITLE
Use pathlib instead of os.path in paths.py

### PR DESCRIPTION
**Pull request by davbrito**
Replacing `os.path` by `pathlib` in `IPython/path.py`. See: #12515

### CODE REVIEW

**Comment by davbrito** [comment]
I am having problem passing [this test](https://github.com/ipython/ipython/blob/3c2c1857268fba1833b324d8b4e4ce53b0220b83/IPython/core/tests/test_paths.py#L57), and others similar.

> ```python
> def test_get_ipython_dir_2():
>     """test_get_ipython_dir_2, Testcase to see if we can call get_ipython_dir without Exceptions."""
>     with patch_get_home_dir('someplace'), \
>             patch.object(paths, 'get_xdg_dir', return_value=None), \
>             patch.object(paths, '_writable_dir', return_value=True), \
>             patch('os.name', "posix"), \
>             modified_env({'IPYTHON_DIR': None,
>                           'IPYTHONDIR': None,
>                           'XDG_CONFIG_HOME': None
>                          }):
>         ipdir = paths.get_ipython_dir()
> 
>     nt.assert_equal(ipdir, os.path.join("someplace", ".ipython"))
> ```

The problem is this line: `patch('os.name', "posix")`. Because pathlib thinks it is running into a unix system then tries to use `PosixPath`.
Therefore it gets this error:
```
======================================================================
ERROR: test_get_ipython_dir_2, Testcase to see if we can call get_ipython_dir without Exceptions.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\David\Envs\ipy\lib\site-packages\nose\case.py", line 198, in runTest
    self.test(*self.arg)
  File "c:\users\david\proyectos\ipython\IPython\core\tests\test_paths.py", line 67, in test_get_ipython_dir_2
    ipdir = paths.get_ipython_dir()
  File "c:\users\david\proyectos\ipython\IPython\paths.py", line 45, in get_ipython_dir
    ipdir = Path(home_dir, ipdir_def)
  File "c:\python\python38\lib\pathlib.py", line 1035, in __new__
    raise NotImplementedError("cannot instantiate %r on your system"
NotImplementedError: cannot instantiate 'PosixPath' on your system
```

**Comment by davbrito** [comment]
@Carreau  This is ready for review

**Comment by davbrito** [comment]
@Carreau this pull request is ready for review

File: .gitignore [hunk 1]
```diff
@@ -28,3 +28,4 @@ __pycache__
 .python-version
 venv*/
 .idea/
+.mypy_cache
```
Comment (Carreau) [inline .gitignore:31]: YOu likely want to put this in your machine global .gitignore as well.

File: IPython/core/traitlets.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+        if isinstance(value, os.PathLike):
+            return value
+        if isinstance(value, str):
+            return Path(value)
+        self.error(obj, value)
+
+    def from_string(self, s):
+        if self.allow_none and s == "None":
+            return None
+        return Path(s)
```
[the remark this answers is not in the document]
Comment (davbrito) [reply IPython/core/traitlets.py:23]: You mean that I open an issue at the [trailets repo](https://github.com/ipython/traitlets)?

**Comment by davbrito** [comment]
> There seem to be a number of commented test with TODO and the rebasing went wrong, I'll try to fix some of this.

I talked about those tests at #12651

**Pull request closed without merging**